### ISSUE TITLE
reland non-regressing changes to self-hosting strings in #996

### ISSUE THREAD

**Issue by mykmelez** [mozilla/j2me.js#1189]
#1171 reverted #996 because it caused #1149 and #1150, but not all of the various changes in #996 caused one of those perf regressions, so we should reland the ones that didn't.

**Comment by mbebenita** [comment]
There are quite a few changes there that we need to reapply. I think it's easier to revert the revert and then work through the performance regressions instead of cherry picking some of the changes in the PR.

**Comment by mykmelez** [comment]
> There are quite a few changes there that we need to reapply. I think it's easier to revert the revert and then work through the performance regressions instead of cherry picking some of the changes in the PR.

We can revert the revert and fix the perf regressions on a branch that we then land, but we shouldn't simply reland the existing set of changes with the perf regressions, because we can't afford them.

**Comment by brendandahl** [comment]
In my strings branch I have rebased and played with this a bit. It seems to helps memory and with a few changes I was able to get the benchmarks closer to the originals, but it still needs some more work.

**Comment by marco-c** [comment]
> In my strings branch I have rebased and played with this a bit. It seems to helps memory and with a few changes I was able to get the benchmarks closer to the originals, but it still needs some more work.

Do we need to re-land all the changes? Couldn't we split them in different PRs?

**Comment by brendandahl** [comment]
Yeah, there's a few pieces that we could split out. Removing the changes to the UTF readers/writers would require us to rewrite them to support the self hosted strings though.
